### ISSUE TITLE
Smart tab jumps the cursor to the end of the file instead of after the closing `)` in a new functions parameters

### ISSUE THREAD

**Issue by David-Else** [helix-editor/helix#7804]
### Summary

When creating a new function smart tab does not tab over the closing parenthesis but jumps to the end of the file.

### Reproduction Steps

Paste this bash code:

```bash
# shellcheck shell=bash

# Adds a line of code to a file if it is not there already
# ${1} file path ${2} line of code to add to file
add_to_file() {
    touch "$1"
    grep -qxF "$2" "$1" && echo "$2 exists in ${GREEN}$1${RESET}" || echo "$2" >>"$1"
}

# Confirm the user is either normal or root and exit if they are not
# ${1} normal|root
confirm_user_is() {
    USER_STATUS=$(id -u)
    if [[ "$USER_STATUS" != 0 && ${1} == "root" ]]; then
        echo "You're not root! Run script with sudo" && exit 1
    elif [[ "$USER_STATUS" == 0 && ${1} == "normal" ]]; then
        echo "You're root! Run script as user" && exit 1
    fi
}

# Detects OS, and if valid sets a global variable OS to valid_rhel/valid_fedora
detect_os() {
    if [[ ("$ID" == "eurolinux" || "$ID" == "centos" || "$ID" == "rocky" || "$ID" == "rhel" || "$ID" == "almalinux") && "${VERSION_ID%.*}" -gt 8 ]]; then
        OS='valid_rhel'
    elif [[ "$ID" == "fedora" && "${VERSION_ID%.*}" -gt 35 ]]; then
        OS='valid_fedora'
    else
        OS='invalid'
    fi
}

# Display a block of text with color ANSI escape codes
display_text() {
    echo -e "$(
        cat <<EOL
        ${1}
EOL
    )"
}
```

Try and create a new function after the first function:

`hello(` it creates the second `)` for you, press `tab` and it jumps to the end of the file.

### Helix log

n/a

### Platform

Linux

### Terminal Emulator

Kitty

### Helix Version

23.05-283-g15e07d4d (latest master)

**Comment by gabydd** [comment]
ah this is because ``hello()`` creates something like
```
(function_definition
<whole rest of the document as the body>)
```
this function definition is the parent of the brackets so move_parent_node_end will move you to the end of the document as the function definition is parsed as to the end of the document

**Comment by dead10ck** [comment]
Yep this is just an issue with how the grammar parses incomplete nodes

**Comment by David-Else** [comment]
Sorry to say it, but smart tabs are totally broken until this is fixed :( 

How does a user know when they can press tab and have it jump over the auto inserted closing pair? It can't work in some situations but "trigger a bug" (as far as the user is concerned) and leap the cursor to the end of the document in others.

I have not really played with this feature much yet, but I found this flaw within 10 seconds of trying it out for the first time, it is going to really throw people.

**Comment by David-Else** [comment]
@gabydd I don't think this should be an upstream issue unless it really is a bug that will be fixed in the future, is that the case?

**Comment by gabydd** [comment]
yeah I guess you are correct, this probably won't be solved upstream, @dead10ck would know better if it is possible to solve in helix though. Also do you have an example of where this happens in a rust file as you mentioned in the issue description?

**Comment by David-Else** [comment]
I just tried to trigger it in Rust again and couldn't, I can't remember exactly what I did. I will remove that reference to Rust and Markdown for now and put it back if I find the problem again. Cheers.

**Comment by dead10ck** [comment]
I mean... this is a general problem with tree-sitter. It can't make a sensible tree to navigate when the syntax isn't valid. It can't guess what you're going to write next. You'll notice that with the same input, running `expand_selection` also selects the whole document. Also, the particular issue of expanding to the whole document only happens if what you're writing is at the very end of the document.

It's an issue that's highly dependent upon the grammar. Some grammars can handle incomplete nodes much better than others.

**Comment by dead10ck** [comment]
That said, it may be sensible to make a separate command that moves the cursor to the end of the surrounding pair. This would not depend upon tree-sitter grammars, but would be strictly less useful in valid syntax trees.

**Comment by dead10ck** [comment]
Also for what it's worth, in my opinion, for this case hitting `()` on the keyboard is always gonna be faster than `(<tab>`.

**Comment by pascalkuthe** [comment]
Yeah I think the command functions as intended here. Jump to parent node end is not the same as the next bracket end and the usecase seems a bit weird conpared to just keep typying in this case.

Altough I could see jumping out tk the next bracket as useful too. That would essentially just binding `mm` to smart tab (and moving one to the right plus not jumping to the start if on the last bracket). That would be pretty easy to implement. The question is just how we would handle the configuration here.

**Comment by CptPotato** [comment]
> this is a general problem with tree-sitter. It can't make a sensible tree to navigate when the syntax isn't valid.

Is it possible to detect when tree sitter's matching fails? In that case we could fall back to something similar to `move_next_word_start`.

**Comment by pascalkuthe** [comment]
No I don't think it's possible to detect that.

**Comment by ModProg** [comment]
I have a reproduction in rust, which is especially annoying when writing doc comments, as it happens whenever you hit `<TAB>` when there is non-whitespace left of the cursor (e.g., inside `///`).

```rs
anything |

```

When I hit tab, I end up in a new line at the end of the file, but only when the file has a trailing new line.

This is one of the reason I'm currently switching back to Neovim whenever I do documentation (though I'll probably keep doing that until helix can preserve `///` when creating a new line and highlight rust code inside the doc comment).

**Comment by andradei** [comment]
I get this sometimes on `lua` scripts and I get that everywhere on `bash` and `fish` scripts.

If would really be great to either:
- Not rely on tree-sitter
- When tree-sitter "fails", fallback to using <kbd>TAB</kbd> as it normally works (insert a tab)

**Comment by stratacast** [comment]
I encountered this now editing an ssh config file. For example

```
example.com
    User<tab>
```

Jumps me to the very end of the file. This makes ssh config files uneditable in Helix

**Comment by heaths** [comment]
I'm also hitting this in a CODEOWNERS file. I'm inserting a line around the middle of the file and, when in INSERT mode, hit `tab` and jumped to the end of the file.

**Comment by dead10ck** [comment]
I'm going to close this, as it's not really actionable. As far as what smart tab is supposed to do, which is jump to the end of the parent node, this is either expected behavior (i.e. the parent node legitimately spans the whole file) or outside the control of Helix (e.g. the grammar doesn't gracefully handle invalid syntax while typing).

For the former case, I think just jumping to the end of the parent node is simply not what you want. I'd recommend binding a different command to tab or just disabling smart tab.
